**Recuerda que una vez abierto, Da clic en "Copiar en Drive", de lo contrario no podras almacenar tu progreso**

Nota: no olvide ir ejecutando las celdas de código de arriba hacia abajo para que no tenga errores de importación de librerías o por falta de definición de variables.

In [1]:
#configuración del laboratorio
# Ejecuta esta celda!
%load_ext autoreload
%autoreload 2
in_colab = True
import os

if not in_colab:
    import sys ; sys.path.append('../commons/utils/');
else: 
    os.system('wget https://raw.githubusercontent.com/mariabda2/ML_2022/master/Labs/commons/utils/general.py -O general.py')
    from general import configure_lab4
    configure_lab4()
from lab4 import *

GRADER = part_1()
sns.set_context('talk')

--2022-07-24 18:45:40--  https://raw.githubusercontent.com/mariabda2/ML_2022/master/Labs/commons/utils/general.py
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.110.133, 185.199.109.133, 185.199.108.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.110.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 15337 (15K) [text/plain]
Saving to: ‘general.py’

     0K .......... ....                                       100% 10.7M=0.001s

2022-07-24 18:45:41 (10.7 MB/s) - ‘general.py’ saved [15337/15337]



lab configuration started
installing libraries
  Using cached pyasn1_modules-0.2.8-py2.py3-none-any.whl (155 kB)
  Using cached requests_oauthlib-1.3.1-py2.py3-none-any.whl (23 kB)
  Using cached pyasn1-0.4.8-py2.py3-none-any.whl (77 kB)
  Using cached oauthlib-3.2.0-py3-none-any.whl (151 kB)


sh: 1: cannot open 3.3.1: No such file


  Using cached scikit_learn-1.1.0-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (31.0 MB)
  Attempting uninstall: scikit-learn
    Found existing installation: scikit-learn 1.1.1
    Uninstalling scikit-learn-1.1.1:
      Successfully uninstalled scikit-learn-1.1.1
downloading files


--2022-07-24 18:45:50--  https://raw.githubusercontent.com/mariabda2/ML_2022/master/Labs/commons/utils/data/AirQuality.data
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.110.133, 185.199.109.133, 185.199.108.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.110.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 567677 (554K) [text/plain]
Saving to: ‘AirQuality.data’

     0K .......... .......... .......... .......... ..........  9% 3.47M 0s
    50K .......... .......... .......... .......... .......... 18% 3.77M 0s
   100K .......... .......... .......... .......... .......... 27% 8.93M 0s
   150K .......... .......... .......... .......... .......... 36% 3.52M 0s
   200K .......... .......... .......... .......... .......... 45% 9.60M 0s
   250K .......... .......... .......... .......... .......... 54% 3.56M 0s
   300K .......... .......... .......... .......... .......... 63% 6.16M 0s
   350K

lab configured


NameError: name 'sns' is not defined

# Laboratorio 4 - Parte 1. Redes neuronales - perceptrón multicapa

Este ejercicio tiene como objetivo implementar una red neuronal artificial de tipo perceptrón multicapa (MLP) para resolver un problema de regresión. Usaremos la librería sklearn. Consulte todo lo relacionado con la definición de hiperparámetros, los métodos para el entrenamiento y la predicción de nuevas muestras en el siguiente enlace: http://scikit-learn.org/stable/modules/generated/sklearn.neural_network.MLPRegressor.html#sklearn.neural_network.MLPRegressor

Para este ejercicio usaremos la base de datos sobre calidad del aire, que ha sido usada en laboratorios previos, pero en este caso trataremos de predecir dos variables en lugar de una, es decir, abordaremos **un problema de múltiples salidas**.

In [ ]:
#cargamos la bd que está en un archivo .data y ahora la podemos manejar de forma matricial
if in_colab:
    db = np.loadtxt('AirQuality.data',delimiter='\t')
else:
    db = np.loadtxt('../commons/utils/data/AirQuality.data',delimiter='\t') 

#Esta es la base de datos AirQuality del UCI Machine Learning Repository. En la siguiente URL se encuentra toda
#la descripción de la base de datos y la contextualización del problema.
#https://archive.ics.uci.edu/ml/datasets/Air+Quality#

x = db[:,0:11]
y = db[:,11:13]

Para calcular los errores, vamos a explorar y usar el [modulo de metricas den sklearn](https://scikit-learn.org/stable/modules/classes.html#module-sklearn.metrics).

Podemos observar que el modulo tiene metricas para regresión y clasificación.

## Ejercicio 1 - Experimentar con MLP para regresión

Para porder implementar nuestra función, lo primero que debemos entender la estrucutra de la red. 

Como mencionamos, vamos a solucionar un problema de multiples salidas. Estas salidas con valores continuos. Por lo tanto debemos garantizar que la capa de salida de nuestra red tenga la capacidad de modelar este tipo de datos.


In [ ]:
#@title Pregunta Abierta
#@markdown ¿De acuerdo al problema planteado, que función de activación debe usar el MLP para un problema de regresión?
respuesta_1 = "" #@param {type:"string"}

Una característica de los modelos de sklearn, es que ciertos tipos de atributos, sólo pueden ser accedidos cuanto se entrena el modelo, explora las siguiente celdas de código:

In [ ]:
mlp = MLPRegressor()
xrandom = np.random.rand(10,2)
yrandom = np.zeros(10)
mlp.fit(X=xrandom, y=yrandom)
print(mlp.activation)
print(mlp.out_activation_)

In [ ]:
mlp = MLPRegressor(activation='logistic')
xrandom = np.random.rand(10,2)
yrandom = np.zeros(10)
mlp.fit(X=xrandom, y=yrandom)
print(mlp.activation)
print(mlp.out_activation_)

In [ ]:
#@title Pregunta Abierta
#@markdown ¿Cual es la diferencia entre las dos funciones de activación?
#@markdown ¿Cual de las opciones está mejor justificada?
 
#@markdown A) `identity` es una función lineal y parte de la capa de entrada. El parámetro `activation` configura la capa de salida.
 
#@markdown B) `identity` es una representación de no tener función activación es parte de la capa de entrada. El parámetro `activation` configura la función de activación de capa de salida.
 
#@markdown C)`identity` es una representación de no tener función activación es parte de la capa de salida.  El parámetro `activation` configura la función de activación de las capas ocultas.
 
#@markdown D) `identity` es una representación de no tener función activación es parte de la capa oculta.  El parámetro `activation` configura la función de activación de la capa de salida.
 
#@markdown Selecciona dentro las lista desplegable
respuesta_2 = '' #@param ["", "A", "B", "C", "D"]

Una vez entendido como sklearn usa las funciones de activación, vamos a crear la función para realizar los experimentos.
 
Vamos a completar la función con el código necesario para usar una red neuronal tipo MLP para solucionar el problema de regresión propuesto.
1. Como función de activación en las capas ocultas use la función 'tanh'.
2. Ajuste el número máximo de épocas a 300.
3. Dejamos como variables el número de capas ocultas y el número de neuronas por capa.
5. debemos seleccionar la función adecuada del [modulo de sklearn para calcular el Error Porcentual Absoluto Medio (MAPE en sigla en ingles)](https://scikit-learn.org/stable/modules/classes.html?highlight=metrics#module-sklearn.metrics). Tener en cuenta qué parámetros usar.
6. Debemos usar los nombres explícitos, por ejemplo si para el MLP es necesario usar el parámetro `activation`, debe ser llamado: `MLPRegressor(activation=...)`
7. Explorar qué hace la siguiente linea de codigo `tuple(2*[100])`
 
**NOTA**: observe el  parámetro `random_state=1` por favor conservarlo, ya que esto hace que los resultados sean similares a lo largo de las ejecuciones.

In [ ]:
# ejercicio de código

def experimetar_mlp(num_hidden_layers, num_neurons, X,Y):
    """ función para realizar experimentos con el MLP
    x: matriz de numpy con caracteristicas
    y: vector numpy con las variables a predecir
    num_hidden_layers: list de enteros con el numero de capdas
        ocultas a usar
    num_neurons: list de enteros con el numero de neuronas a usar
    
    Retorna: dataframe con 6 columnas:
        - numero de capas, numero de neuronas
        - promedio de error prueba variable 1 y desviación estandar
        - promedio de error prueba variable 2 y desviación estandar
        
    """
    #Validamos el modelo
    Folds = 3
    ss = ShuffleSplit(n_splits=Folds, test_size=0.2, random_state = 1)
    resultados = pd.DataFrame()
    idx = 0
    for hidden_layers in num_hidden_layers:
        for neurons in num_neurons:
            for j, (train, test) in enumerate(ss.split(X)):
                # para almacenar errores intermedios
                ErrorY1 = np.zeros(Folds)
                ErrorY2 = np.zeros(Folds)
                Xtrain= X[train,:]
                Ytrain = Y[train,:]
                Xtest = X[test,:]
                Ytest = Y[test,:]
                #Normalizamos los datos
                scaler = StandardScaler().fit(...)       
                Xtrain = scaler.transform(...)
                Xtest = scaler.transform(...)
                #Haga el llamado a la función para crear y entrenar el modelo usando los datos de entrenamiento
                # prestar atención a los parametros, correctos.
                hidden_layer_sizes = tuple(...)
                mlp = MLPRegressor(..., random_state=1)
                mlp...(....)
                #Use para el modelo para hacer predicciones sobre el conjunto Xtest
                Yest = mlp...(...)
                #Mida el error absoluto medio para cada una de las dos salidas
                #Observe bien la documentación. recordar que esta resolviendo
                #un problema de multiples salidas
                errors = ....(y_true=..., y_pred = ..., multioutput=...)
                ErrorY1[j] = errors[0]
                ErrorY2[j] = errors[1]
        
            print('error para salida 1 = ' + str(np.mean(ErrorY1)) + '+-' + str(np.std(ErrorY1)))
            print('error para 2 = ' + str(np.mean(ErrorY2)) + '+-' + str(np.std(ErrorY2)))
        
            resultados.loc[idx,'capas ocultas'] = hidden_layers
            resultados.loc[idx,'neuronas en capas ocultas'] = neurons 
            resultados.loc[idx,'error de prueba y1(media)'] = ...(ErrorY1)
            resultados.loc[idx,'intervalo de confianza y1'] = ...(ErrorY1)
            resultados.loc[idx,'error de prueba y2(media)'] = ...(ErrorY2)
            resultados.loc[idx,'intervalo de confianza y2'] = ...(ErrorY2)
            idx+=1
    return (resultados)

In [ ]:
## la funcion que prueba tu implementacion
# ignorar los prints
GRADER.run_test("ejercicio1", experimetar_mlp)

vamos a realizar los experimentos

In [ ]:
# tarda unos minutos!!
resultados_mlpr = experimetar_mlp(num_hidden_layers = [1,2,3], num_neurons  = [8,12,16], X=x, Y=y)

In [ ]:
sns.relplot(data = resultados_mlpr,  x='neuronas en capas ocultas', y = 'error de prueba y1(media)', style= 'capas ocultas', kind = 'line')

In [ ]:
#@title Pregunta Abierta
#@markdown Con base a los reusltados, ¿podriamos saber a priori los mejores parametros (capas ocultas, # de neuronas), que tan correcto es asumir patrones en los valores de los parametros?
respuesta_3 = "" #@param {type:"string"}

In [ ]:
#@title Pregunta Abierta
#@markdown ¿Cual es el rol de random_state en la función `MLPRegressor`?
#@markdown ¿Cual de las opciones está mejor justificada?
 
#@markdown A) Usando los valores por defecto, determina la inicialización de los pesos de cada una de las neuronas en cada capa del perceptrón.
 
#@markdown B) Sin importar los otros valores, determina SOLO la inicialización de los pesos de cada una de las neuronas en cada capa del perceptrón.
 
#@markdown C) Determina la inicialización y las funciones de activación de los pesos de cada una de las neuronas en cada capa del perceptrón.
 
 #@markdown D) Usando los valores por defecto, determina la inicialización de los pesos de la capa de salida del perceptrón.
 
#@markdown Selecciona dentro las lista desplegable
respuesta_4 = '' #@param ["", "A", "B", "C", "D"]



## Ejercicio 2 Experimentar con MLP para clasificación

A continuación se leen los datos de un problema de clasificación. El problema corresponde a la clasifiación de dígitos escritos a mano. Usaremos únicamente 4 de las 10 clases disponibles. Los datos fueron preprocesados para reducir el número de características. La técnica usada será analizada más adelante en el curso.

In [ ]:
digits = load_digits(n_class=4)
#--------- preprocesamiento--------------------
pca = PCA(0.99, whiten=True)
data = pca.fit_transform(digits.data)
#---------- Datos a usar ----------------------
Xd = data
Yd = digits.target

In [ ]:
#@title Pregunta Abierta
#@markdown  ¿Qué tipo de función de activación usa el modelo en la capa de salida para un problema de clasificación?
respuesta_5 = "" #@param {type:"string"}

como lo hicmos antes, vamos a comprobar con la libreria la función de activación

In [ ]:
# ejercicio de código
def output_activation_MPC():
    """funcion que entrena un modelo
    con data aleatoria para confirmar la funcion
    de activacion de la ultima capa
    """
    mlp = ...()
    # fit with some random data
    xrandom = np.random.rand(10,2)
    yrandom = np.zeros(10)
    # llamar el metodo adecuado para entrenar
    # el mlp con los x y 'y' random
    mlp.fit(...)
    # retornar el atributo de mlp adecuado
    return (...)

In [ ]:
## la funcion que prueba tu implementacion
GRADER.run_test("ejercicio2", output_activation_MPC)

In [ ]:
print("la función de activación para un problema de clasificación es:", output_activation_MPC())

Ahora en nuestro siguiente ejercicio vamos a implementar una red neuronal artificial de tipo perceptrón multicapa (MLP) para resolver un problema de clasificación. Usaremos la librería sklearn. Consulte todo lo relacionado con la definición de hiperparámetros, los métodos para el entrenamiento y la predicción de nuevas muestras en el siguiente enlace: http://scikit-learn.org/stable/modules/generated/sklearn.neural_network.MLPClassifier.html#sklearn.neural_network.MLPClassifier


Vamos completar la función con el código necesario para usar una red neuronal tipo MLP para solucionar el problema de clasificación propuesto.
1. Como función de activación en las capas ocultas use la función tangencial hiperbólica. 
2. Ajuste el número máximo de épocas a 350.
3. Dejamos como variables el número de capas ocultas y el número de neuronas por capa
5. Seleccione la función adecuada del [modulo de sklearn para calcular la exactitud del clasificador](https://scikit-learn.org/stable/modules/classes.html?highlight=metrics#module-sklearn.metrics). Tener en cuenta que parametros usar.
6. Debemos usar los nombres explicitos, por ejemplo si para el MLP es necesario usar el parametro `activation`, debe ser llamado: `MLPClassifier(activation=...)`

**NOTA**: cuando observe el el parametro `random_state=1` por favor conservarlo, ya que esto hace que los resultados sean similares a lo largo de las ejecucciones. *tener en cuenta esta observación para una de las preguntas abiertas que encuentras mas adelante del laboratorio.

In [ ]:
# ejercicio de código
def experimetar_mlpc(X,Y, num_hidden_layers, num_neurons):
    """ función para realizar experimentos con el MLP
    x: matriz de numpy con caracteristicas
    y: vector numpy con las variables a predecir
    num_hidden_layers: list de enteros con el numero de capdas
        ocultas a usar
    num_neurons: list de enteros con el numero de neuronas a usar
    
    Retorna: dataframe con 4 columnas:
        - numero de capas, numero de neuronas
        - promedio de error prueba (exactitud/eficiencia) de claisficacion y desviación estandar        
    """
    #Validamos el modelo
    Folds = 3
    skf = StratifiedKFold(n_splits=Folds)
    resultados = pd.DataFrame()
    idx = 0
    for hidden_layers in num_hidden_layers:
        for neurons in num_neurons:
            for j, (train, test) in enumerate(skf.split(..., ...)):
                # para almacenar errores intermedios
                Error = np.zeros(Folds)
                Xtrain= X[train,:]
                Ytrain = Y[train]
                Xtest = X[test, :]
                Ytest = Y[test]
                #Normalizamos los datos
                scaler = StandardScaler().fit(Xtrain)       
                Xtrain = scaler.transform(Xtrain)
                Xtest = scaler.transform(Xtest)
                #Haga el llamado a la función para crear y entrenar el modelo usando los datos de entrenamiento
                # prestar atención a los parametros, correctos.
                hidden_layer_sizes = tuple(hidden_layers*[neurons])
                mlp = MLPClassifier(..., random_state = 1)
                # entrenar el MLP
                ..fit(....)
                #Use para el modelo para hacer predicciones sobre el conjunto Xtest
                Yest = ...predict(...)
                # recordar usar la medida adecuada de acuerdo a las instrucciones
                Error[j] = ...(y_true = ...,  y_pred =...)
                
        
            print('error para configuracion de params = ' + str(np.mean(Error)) + '+-' + str(np.std(Error)))
        
            resultados.loc[idx,'capas ocultas'] = hidden_layers
            resultados.loc[idx,'neuronas en capas ocultas'] = neurons 
            resultados.loc[idx,'error de prueba(media)'] = np.mean(Error)
            resultados.loc[idx,'intervalo de confianza'] = np.std(Error)
            idx+=1
    return (resultados)

In [ ]:
## la funcion que prueba tu implementacion
GRADER.run_test("ejercicio3", experimetar_mlpc)

In [ ]:
# tarda unos minutos!!
resultados_mlpc = experimetar_mlpc(X = Xd, Y=Yd, num_hidden_layers=[1,2,3], num_neurons=[12,16,20])

In [ ]:
resultados_mlpc

In [ ]:
# ver los resultados
# notar como las capas ocultas y el # de neuronas influyen
sns.relplot(data = resultados_mlpc,  x='neuronas en capas ocultas', y = 'error de prueba(media)', style= 'capas ocultas', kind = 'line')

In [ ]:
#@title Pregunta Abierta
#@markdown ¿Cuántas neuronas en la capa de salida tiene el modelo? ¿Porqué debe tener ese número?
respuesta_6 = "" #@param {type:"string"}

In [ ]:
#@title Pregunta Abierta
#@markdown Llegas a una nueva compañía. En todos los proyectos observas que solo tienen perceptrones multicapa entrenados usado sklearn ¿que recomiendas?
respuesta_7 = "" #@param {type:"string"}

**recordatorio** En la practica sklearn no es una la libreria indicada para desarollar redes neuronales, para practicas mas avanzadas y realizar modelos en el "mundo real" [se deben usar conceptos de deep learning y una libreria llamada Keras](https://colab.research.google.com/github/lexfridman/mit-deep-learning/blob/master/tutorial_deep_learning_basics/deep_learning_basics.ipynb). Adicional tener en cuenta [lo visto en teoria](https://mariabda2.github.io/ML_2022/clases/clase_14_Redes_Neuronales_Artificiales.html#desventajas-de-las-aproximaciones-clasicas).

In [ ]:
GRADER.check_tests()

In [ ]:
#@title Integrantes
codigo_integrante_1 ='' #@param {type:"string"}
codigo_integrante_2 = ''  #@param {type:"string"}

----
esta linea de codigo va fallar, es de uso exclusivo de los profesores


In [ ]:
GRADER.grade()